libcalico-go/lib/backend/syncersv1/dedupebuffer/dedupe_buffer.go

```diff
@@ -0,0 +1,283 @@
+			if err != nil {
+				// Shouldn't happen, we get out keys from Typha which has already
+				// Encoding them once!
+				log.WithError(err).WithField("key", u.Key).Error(
+					"Failed to generate default path for key.  Will skip this update.")
+				continue
+			}
+		}
+
+		if element, ok := d.keyToPendingUpdate[key]; ok {
+			// Already got an in-flight update for this key.
+			if u.Value == nil && !d.sentKeys.Contains(key) {
+				// This is a deletion, but the key in question never made it
+				// off the queue, remove it entirely.
+				if debug {
+					log.WithField("key", key).Debug("Key deleted before being sent.")
+				}
+				delete(d.keyToPendingUpdate, key)
+				d.pendingUpdates.Remove(element)
+			} else {
+				// Update to a key that's already on the queue, swap in the
+				// most recent value.
+				if debug {
+					log.WithField("key", key).Debug("Key updated before being sent.")
+				}
+				usk := element.Value.(updateWithStringKey)
+				usk.update = u
+				element.Value = usk
+			}
+		} else {
+			// No in-flight entry for this key.  Add to queue and record that
+			// it's in flight.
+			if debug {
+				log.WithField("key", key).Debug("No in flight value for key, adding to queue.")
+			}
+			element = d.pendingUpdates.PushBack(updateWithStringKey{
+				key:    key,
+				update: u,
+			})
+			d.keyToPendingUpdate[key] = element
+		}
+	}
+	queueNowEmpty := d.pendingUpdates.Len() == 0
+	if queueWasEmpty && !queueNowEmpty {
+		// Only need to signal when the first item goes on the queue.
+		if debug {
+			log.Debug("Queue transitioned to non-empty; signalling.")
+		}
+		d.cond.Signal()
+	}
+}
+
+func (d *DedupeBuffer) SendToSinkForever(sink api.SyncerCallbacks) {
+	for !d.stopped {
+		d.sendNextBatchToSink(sink)
+	}
+}
+
+func (d *DedupeBuffer) Stop() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.stopped = true
+	d.cond.Signal()
+}
+
+type updateWithStringKey struct {
+	key    string
+	update api.Update
+}
+
+func (d *DedupeBuffer) sendNextBatchToSink(sink api.SyncerCallbacks) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	for d.pendingUpdates.Len() == 0 {
+		if d.stopped {
+			return
+		}
+		d.cond.Wait()
+	}
+	const batchSize = 100
+	buf := make([]any, 0, batchSize)
+	for d.pendingUpdates.Len() > 0 {
+		if d.stopped {
+			return
+		}
+
+		buf = d.pullNextBatch(buf, batchSize)
+		d.dropLockAndSendBatch(sink, buf)
+	}
+}
+
+func (d *DedupeBuffer) pullNextBatch(buf []any, batchSize int) []any {
+	// Grab a batch of updates off the queue.
+	buf = buf[:0]
+	for len(buf) < batchSize && d.pendingUpdates.Len() > 0 {
+		first := d.pendingUpdates.Front()
+		buf = append(buf, first.Value)
+		d.pendingUpdates.Remove(first)
+		if u, ok := first.Value.(updateWithStringKey); ok {
+			key := u.key
+			delete(d.keyToPendingUpdate, key)
+			// Update sentKeys now, before we drop the lock.  Once we drop
+			// the lock we're committed to sending these keys.
+			if u.update.Value == nil {
+				d.sentKeys.Discard(key)
```

Comment:
Perhaps `liveResourceKeys` would be a better name than `sentKeys`?  With `sentKeys` my intuition was that we would never discard from the set.  The coding this way is certainly better, but then the naming is a little misleading.